### DISCUSSION TITLE
Theory with Type for argument unable to resolve

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [xunit/xunit#2508] [Feedback Request]
Maybe this is more of an issue, I do not know. I have a `Theory` something like this:

```csharp
[Theory
    , InlineData(typeof(Path.To.Root))
    ]
public virtual void Verify_Root_Type(Type rootType)
{
    // Do something to assert rootType, etc...
}
```

Where:

```csharp
namespace Path.To
{
    public class Root
    {
        // ...
    }
}
```

However, it seems that in `VS2022` at any rate discovery never finishes.

Is there a better way to convey this, maybe I need a wrapper that has a better stringification?

Thank you...

**Comment by cihatsolak** [comment]
The issue with the provided Theory may be due to the fact that Type is not a serializable type, and therefore may not be properly handled during the test discovery process. One possible workaround could be to instead pass the fully-qualified name of the type as a string, and then use reflection to obtain the actual Type object within the test method. For example:

```
[Theory]
[InlineData("Path.To.Root")]
public virtual void Verify_Root_Type(string rootTypeName)
{
  Type rootType = Type.GetType(rootTypeName);
  // Do something to assert rootType, etc...
}
```

Alternatively, a custom TypeConverter could be used to convert the string representation of the type to the actual Type object.

**Reply by bradwilson** [reply]
Type is generally serializable, with some limitations.

Serialize:

https://github.com/xunit/xunit/blob/1892224c60f8499b42ee1d8a521d65995ef32aea/src/common/XunitSerializationInfo.cs#L309-L311

https://github.com/xunit/xunit/blob/1892224c60f8499b42ee1d8a521d65995ef32aea/src/common/SerializationHelper.cs#L192-L248

Deserialize:

https://github.com/xunit/xunit/blob/1892224c60f8499b42ee1d8a521d65995ef32aea/src/common/XunitSerializationInfo.cs#L174-L175

https://github.com/xunit/xunit/blob/1892224c60f8499b42ee1d8a521d65995ef32aea/src/common/SerializationHelper.cs#L150-L190

However, if the assembly that the Type references is for some reason not loadable during deserialization, then that could cause a failure.

**Reply by bradwilson** [reply]
I will also reiterate that "source based discovery" is not written by nor maintained by the xUnit.net team. If you are having problems with discovery, the first thing we will recommend you do is disable source based discovery, as shown here:

![image](https://user-images.githubusercontent.com/16944/227093290-c9f3c75d-b8e4-4dd0-bedc-f73909fc4aca.png)

Make sure the indicated check box is unchecked.

Unchecking this box means discovery will only take place as a result of you building your source code. That discovery runs our code, and is the only discovery mechanism we support. If you are having issues that are fixed by disabling this (or fixed simply by building your source code), then the Visual Studio team will be your next stop to get those issues resolved.

**Comment by mwpowellhtx** [comment]
Serializable in terms of what exactly? I'm not sure I understand. `System.Type` itself is serializable, isn't it? Is that not sufficient?

Been a little while since I noticed this, but IIRC, I am using `Root` as an anchor type in order to do some use case specific unit tests when they are overridden.

So therefore it is there as nothing more than an anchor, placeholder, key, along these lines.

**Comment by mwpowellhtx** [comment]
Thanks, I'll have a gander at those options as well.
